extensions/amp-iframe/0.1/amp-iframe.js

```diff
@@ -177,104 +184,136 @@ export class AmpIframe extends AMP.BaseElement {
-      // Chrome does not reflect the iframe readystate.
-      iframe.readyState = 'complete';
-      this.activateIframe_();
-      if (isTracking) {
-        timer.promise(trackingIframeTimeout).then(() => {
-          if (iframe.parentNode) {
-            iframe.parentNode.removeChild(iframe);
-          }
-          this.element.setAttribute('amp-removed', '');
-          this.iframe_ = null;
-          this.iframeSrc = null;
-        });
+      if (!isTracking) {
+        /** @private {!IntersectionObserver} */
+        this.intersectionObserver_ = new IntersectionObserver(this, iframe_);
       }
-    };
 
-    listen(iframe, 'embed-size', data => {
-      if (data.width !== undefined) {
-        iframe.width = data.width;
-        this.element.setAttribute('width', data.width);
+      setSandbox(this.element, iframe, this.sandbox_);
+
+      iframe.src = this.iframeSrc;
+      this.element.appendChild(makeIOsScrollable(this.element, iframe));
+
+      iframe.onload = () => {
+        // Chrome does not reflect the iframe readystate.
+        iframe.readyState = iframe.getAttribute('src') ? 'complete' : '';
+
+        this.activateIframe_();
+
+        if (isTracking) {
+          timer.promise(trackingIframeTimeout).then(() => {
+            if (iframe.parentNode) {
+              iframe.parentNode.removeChild(iframe);
+            }
+            // TODO(jridgewell): vsync.mutate?
+            this.element.setAttribute('amp-removed', '');
+            this.iframe_ = null;
+            this.iframeSrc = null;
+          });
+        }
+      };
+
+      listen(iframe, 'embed-size', data => {
+        if (data.width !== undefined) {
+          iframe.width = data.width;
+          this.element.setAttribute('width', data.width);
+        }
+        if (data.height !== undefined) {
+          const newHeight = Math.max(this.element./*OK*/offsetHeight +
+              data.height - this.iframe_./*OK*/offsetHeight, data.height);
+          iframe.height = data.height;
+          this.element.setAttribute('height', newHeight);
+          this.updateHeight_(newHeight);
+        }
+      });
+
+      if (this.isClickToPlay_) {
+        listen(iframe, 'embed-ready', this.activateIframe_.bind(this));
       }
-      if (data.height !== undefined) {
-        const newHeight = Math.max(this.element./*OK*/offsetHeight +
-            data.height - this.iframe_./*OK*/offsetHeight, data.height);
-        iframe.height = data.height;
-        this.element.setAttribute('height', newHeight);
-        this.updateHeight_(newHeight);
-      }
-    });
-    if (this.isClickToPlay_) {
-      listen(iframe, 'embed-ready', this.activateIframe_.bind(this));
+
+      this.iframe_ = iframe;
+    }
+
+    if (iframe.getAttribute('src') !== this.iframeSrc) {
+      iframe.src = this.iframeSrc;
     }
+
     return loadPromise(iframe);
   }
 
+
+  /** @override */
+  documentInactiveCallback() {
+    if (this.iframe_) {
+      this.iframe_.src = '';
```

Comment:
It would be better to remove the element from the DOM. It is even OK to keep it in memory. For legacy reason the window is actually destroyed when the iframe is removed from the DOM and so the major memory usage goes away.